### ISSUE TITLE
KNearestNeighbors Constructor (Int32, Int32, Double[][], Int32[], IMetric(Double[]))

### ISSUE THREAD

**Issue by Lyubna** [accord-net/framework#386]
How to generate the IMetric(Double[]) parameter? I need to calculate the Cosine distance 

My input is double[][] 
output is int [] 

it works fine when I use 
                KNearestNeighbors knn = new KNearestNeighbors(k: 10, classes: 3, inputs: inputs, outputs: outputs);

but when I want to change the distance as you showed in the example like that
KNearestNeighbors knn = new KNearestNeighbors(k: 10, classes: 3, inputs: inputs, outputs: outputs,distance: Distance.Cosine);

it shows me the following error: 
Error	1	The best overloaded method match for 'Accord.MachineLearning.KNearestNeighbors.KNearestNeighbors(int, int, double[][], int[], Accord.Math.Distances.IMetric<double[]>)' has some invalid arguments	

Error	2	Argument 5: cannot convert from 'method group' to 'Accord.Math.Distances.IMetric<double[]>'

What can I do to solve this and change the default Euclidean distance to Cosine? 
p.s. I also tried the example you used with classifying the strings, it also showed error

**Comment by cesarsouza** [comment]
Hi there,

Please, can you try creating a new Cosine distance using the new operator, i.e.:
```c#
var knn = new KNearestNeighbors(k: 10, classes: 3, inputs: inputs, outputs: outputs, distance: new Cosine());
```

The [Cosine class resides in the Accord.Math.Distances namespace](http://accord-framework.net/docs/html/T_Accord_Math_Distances_Cosine.htm).

Hope it helps!

Best regards,
Cesar

**Comment by Lyubna** [comment]
Tried that too but it didn't work

**Comment by cesarsouza** [comment]
Oh yeah, it's because K-Nearest Neighbors is expecting a proper distance metric, while the Cosine gives only a similarity distance. 

While I work in the issue, a quick solution is to implement a class that implements the IMetric interface and implement the "Distance" method using [the definition for the Cosine distance you would like to use](https://en.wikipedia.org/wiki/Cosine_similarity#Angular_distance_and_similarity).

Is there any particular paper or model you would like to implement? If yes then I can prepare the proper metric for the model you need.

Regards,
Cesar

**Comment by Lyubna** [comment]
Thanks Cesar, I guess I wrote the formulas right, maybe you can tell me if I'm wrong or right and you can use them as a solution 

`public double Similarity(double[][] x, double[][] y)
        {
            double sum = 0;
            double p = 0;
            double q = 0;
           
            for (int i = 0; i < x.Length && i< y.Length; i++)
            {
                for (int j = 0; j < x.GetLength(1) && j < y.GetLength(1); j++)
                {
                    sum += x[i][j] * y[j][i];
                    p += Math.Pow(x[i][j],2);
                    q += Math.Pow(y[i][j],2);
                }
            }    

                       

            double den = Math.Sqrt(p) * Math.Sqrt(q);
            return (sum == 0) ? 0 : sum / den;
        }
`


`public double Distance(double[][]x, double[][] y)
        {
            return (Math.Acos(Similarity(x, y))/Math.PI);
            
        }`


And thank you so much again for helping me :D

**Comment by cesarsouza** [comment]
Hi there,

Thanks!

Though I have to say there is a small problem in the code you posted related to the way you are getting the lengths of each individual feature vector/point in the inner loop of your Similarity method.

In a nutshell, you might be able to address the problem you were currently facing by implementing the following class:


```c#
[Serializable]
public CosineDistance : IMetric<double[]>
{
    public double Similarity(double[][] x, double[][] y)
    {
        double sum = 0;
        double p = 0;
        double q = 0;

        for (int i = 0; i < x.Length && i< y.Length; i++)
        {
            for (int j = 0; j < x[i].Length && j < y[i].Length; j++)
            {
                sum += x[i][j] * y[j][i];
                p += Math.Pow(x[i][j],2);
                q += Math.Pow(y[i][j],2);
            }
        }    

        double den = Math.Sqrt(p) * Math.Sqrt(q);
        return (sum == 0) ? 0 : sum / den;
    }

    public double Distance(double[][]x, double[][] y)
    {
        return (Math.Acos(Similarity(x, y))/Math.PI);
    }
}
```

And then passing an instance of this class to the kNN constructor you initially referenced.

Hope it helps!

Best regards,
Cesar

**Comment by cesarsouza** [comment]
Fixed in 3.5.0.
